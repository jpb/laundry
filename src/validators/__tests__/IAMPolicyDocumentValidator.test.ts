import { lint } from '../../index';

describe('IAMPolicyDocumentValidator', () => {
  test('valid', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Sid: '',
                    Action: '',
                    Effect: '',
                    Principal: '',
                    Resource: '',
                    Condition: {}
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toEqual([]);
  });

  test('valid, with lists', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Sid: '',
                    Action: '',
                    Effect: '',
                    Principal: [''],
                    Resource: [''],
                    Condition: {}
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toEqual([]);
  });

  test('missing required properties', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {}
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toMatchSnapshot();
  });

  test('invalid types', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Sid: [],
                    Action: [],
                    Effect: [],
                    Principal: {},
                    Resource: {},
                    Condition: '',
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toMatchSnapshot();
  });
});
