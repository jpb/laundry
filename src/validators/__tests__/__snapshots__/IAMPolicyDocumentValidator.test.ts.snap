// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IAMPolicyDocumentValidator invalid effect 1`] = `
Array [
  Object {
    "message": "must be one of Allow, Deny, got \\"foo\\"",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Effect",
    ],
    "source": "IAMPolicyDocumentValidator",
  },
]
`;

exports[`IAMPolicyDocumentValidator invalid types 1`] = `
Array [
  Object {
    "message": "must be a String, got []",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Sid",
    ],
    "source": "IAMPolicyDocumentValidator",
  },
  Object {
    "message": "must be a String, got []",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Effect",
    ],
    "source": "IAMPolicyDocumentValidator",
  },
  Object {
    "message": "must be a String, got {} or must be a List, got {}",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Principal",
    ],
    "source": "IAMPolicyDocumentValidator",
  },
  Object {
    "message": "must be a String, got {} or must be a List, got {}",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Resource",
    ],
    "source": "IAMPolicyDocumentValidator",
  },
  Object {
    "message": "must be an Object, got \\"\\"",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Condition",
    ],
    "source": "IAMPolicyDocumentValidator",
  },
]
`;

exports[`IAMPolicyDocumentValidator missing required properties 1`] = `
Array [
  Object {
    "message": "is required",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Effect",
    ],
    "source": "IAMPolicyDocumentValidator",
  },
  Object {
    "message": "is required",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Action",
    ],
    "source": "IAMPolicyDocumentValidator",
  },
]
`;
